File: snippets/fsharp/System/TimeZoneInfo/BaseUtcOffset/TimeZone2_Examples.fs
module TimeZone2

open System

do
    // <Snippet1>
    let localZone = TimeZoneInfo.Local
    printfn $"""The {localZone.DisplayName} time zone is {abs localZone.BaseUtcOffset.Hours}:{abs localZone.BaseUtcOffset.Minutes} {if localZone.BaseUtcOffset >= TimeSpan.Zero then "later" else "earlier"} than Coordinated Universal Time."""
    // </Snippet1>                  

do
    let localZone = TimeZoneInfo.Local
    printfn $"Local Time Zone ID: {localZone.Id}"
    printfn $"   Display Name is: {localZone.DisplayName}."
    printfn $"   Standard name is: {localZone.StandardName}."
    printfn $"   Daylight saving name is: {localZone.DaylightName}."

do
    // <Snippet3>
    let universalZone = TimeZoneInfo.Utc
    printfn $"The universal time zone is {universalZone.DisplayName}."
    printfn $"Its standard name is {universalZone.StandardName}."
    printfn $"Its daylight savings name is {universalZone.DaylightName}."
    // </Snippet3>

do
    // <Snippet4>
    let zones = TimeZoneInfo.GetSystemTimeZones()
    for zone in zones do
        if not zone.SupportsDaylightSavingTime then
            Console.WriteLine zone.DisplayName
    // </Snippet4> 

do
    // <Snippet5>
    let zones = TimeZoneInfo.GetSystemTimeZones()
    printfn $"The local system has the following {zones.Count} time zones"
    for zone in zones do
        printfn $"{zone.Id}"
    // </Snippet5>

do
    // <Snippet7> 
    let thisTimeZone = TimeZoneInfo.Local
    let zone1 = TimeZoneInfo.FindSystemTimeZoneById "Pacific Standard Time"
    let zone2 = TimeZoneInfo.FindSystemTimeZoneById "Eastern Standard Time"
    printfn $"{thisTimeZone.Equals zone1}"
    printfn $"{thisTimeZone.Equals zone2}"
    // </Snippet7>

do
    // <Snippet8>
    // Specify DateTimeKind in Date constructor
    let baseTime = DateTime(2007, 11, 4, 0, 59, 00, DateTimeKind.Unspecified)

    // Get Pacific Standard Time zone
    let pstZone = TimeZoneInfo.FindSystemTimeZoneById "Pacific Standard Time"

    // List possible ambiguous times for 63-minute interval, from 12:59 AM to 2:01 AM 
    for i = 0 to 62 do
    // Because of assignment, newTime.Kind is also DateTimeKind.Unspecified
        let newTime = baseTime.AddMinutes i   
        printfn $"{newTime} is ambiguous: {pstZone.IsAmbiguousTime newTime}"
    // </Snippet8>      

do
    // <Snippet9>
    // Specify DateTimeKind in Date constructor
    let baseTime = DateTime(2007, 3, 11, 1, 59, 0, DateTimeKind.Unspecified)
    
    // Get Pacific Standard Time zone
    let pstZone = TimeZoneInfo.FindSystemTimeZoneById "Pacific Standard Time"

    // List possible invalid times for a 63-minute interval, from 1:59 AM to 3:01 AM
    for i = 0 to 62 do
        // Because of assignment, newTime.Kind is also DateTimeKind.Unspecified
        let newTime = baseTime.AddMinutes i
        printfn $"{newTime} is invalid: {pstZone.IsInvalidTime newTime}"
    // </Snippet9>

File: snippets/csharp/System/TimeZoneInfo/BaseUtcOffset/TimeZone2_Examples.cs
using System;
using System.Collections.ObjectModel;
using System.Windows.Forms;

[assembly: CLSCompliant(true)]

namespace TimeZoneExamples
{
    public class TZClass
    {
        public static void Run()
        {
            TZClass tz = new();
            if (MessageBox.Show("Display time zone offset?", "Offset", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowTimezoneOffset();

            if (MessageBox.Show("Display time zone names?", "Names", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowTimeZoneNames();

            if (MessageBox.Show("Display universal time zone names?", "Universal Time Zone Names", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowUniversalTimeZoneNames();

            if (MessageBox.Show("Show time zones without daylight savings time", "Zones Supporting DST", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowNoDSTZones();

            if (MessageBox.Show("List all time zone IDs?", "IDs", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowTimeZoneIDs();

            if (MessageBox.Show("Test Time Zones for Equality?", "TimeZoneInfo.Equals", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.TestForEquality();

            if (MessageBox.Show("Show ambiguous times in Pacific Time Zone for 2007?", "TimeZoneInfo.IsAmbiguousTime", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowAmbiguousTimes();

            if (MessageBox.Show("Show invalid times in Pacific Time Zone for 2007?", "TimeZoneInfo.IsInvalidTime", MessageBoxButtons.YesNo) == DialogResult.Yes)
                tz.ShowInvalidTimes();
        }

        private void ShowTimezoneOffset()
        {
            // <Snippet1>
            TimeZoneInfo localZone = TimeZoneInfo.Local;
            Console.WriteLine($"The {localZone.DisplayName} time zone is {Math.Abs(localZone.BaseUtcOffset.Hours)}:{Math.Abs(localZone.BaseUtcOffset.Minutes)} {((localZone.BaseUtcOffset >= TimeSpan.Zero) ? "later" : "earlier")} than Coordinated Universal Time.");
            // </Snippet1>
        }

        private void ShowTimeZoneNames()
        {
            TimeZoneInfo localZone = TimeZoneInfo.Local;
            Console.WriteLine($"Local Time Zone ID: {localZone.Id}");
            Console.WriteLine($"   Display Name is: {localZone.DisplayName}.");
            Console.WriteLine($"   Standard name is: {localZone.StandardName}.");
            Console.WriteLine($"   Daylight saving name is: {localZone.DaylightName}.");
        }

        private void ShowUniversalTimeZoneNames()
        {
            // <Snippet3>
            TimeZoneInfo universalZone = TimeZoneInfo.Utc;
            Console.WriteLine($"The universal time zone is {universalZone.DisplayName}.");
            Console.WriteLine($"Its standard name is {universalZone.StandardName}.");
            Console.WriteLine($"Its daylight savings name is {universalZone.DaylightName}.");
            // </Snippet3>
        }

        private void ShowNoDSTZones()
        {
            // <Snippet4>
            ReadOnlyCollection<TimeZoneInfo> zones = TimeZoneInfo.GetSystemTimeZones();
            foreach (TimeZoneInfo zone in zones)
            {
                if (!zone.SupportsDaylightSavingTime)
                    Console.WriteLine(zone.DisplayName);
            }
            // </Snippet4>
        }

        private void ShowTimeZoneIDs()
        {
            // <Snippet5>
            ReadOnlyCollection<TimeZoneInfo> zones = TimeZoneInfo.GetSystemTimeZones();
            Console.WriteLine($"The local system has the following {zones.Count} time zones");
            foreach (TimeZoneInfo zone in zones)
                Console.WriteLine(zone.Id);
            // </Snippet5>
        }

        private void TestForEquality()
        {
            // <Snippet7>
            TimeZoneInfo thisTimeZone, zone1, zone2;

            thisTimeZone = TimeZoneInfo.Local;
            zone1 = TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time");
            zone2 = TimeZoneInfo.FindSystemTimeZoneById("Eastern Standard Time");
            Console.WriteLine(thisTimeZone.Equals(zone1));
            Console.WriteLine(thisTimeZone.Equals(zone2));
            // </Snippet7>
        }

        private void ShowAmbiguousTimes()
        {
            // <Snippet8>
            // Specify DateTimeKind in Date constructor
            DateTime baseTime = new(2007, 11, 4, 0, 59, 00, DateTimeKind.Unspecified);
            DateTime newTime;

            // Get Pacific Standard Time zone
            TimeZoneInfo pstZone = TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time");

            // List possible ambiguous times for 63-minute interval, from 12:59 AM to 2:01 AM
            for (int ctr = 0; ctr < 63; ctr++)
            {
                // Because of assignment, newTime.Kind is also DateTimeKind.Unspecified
                newTime = baseTime.AddMinutes(ctr);
                Console.WriteLine($"{newTime} is ambiguous: {pstZone.IsAmbiguousTime(newTime)}");
            }
            // </Snippet8>
        }

        private void ShowInvalidTimes()
        {
            // <Snippet9>
            // Specify DateTimeKind in Date constructor
            DateTime baseTime = new(2007, 3, 11, 1, 59, 0, DateTimeKind.Unspecified);
            DateTime newTime;

            // Get Pacific Standard Time zone
            TimeZoneInfo pstZone = TimeZoneInfo.FindSystemTimeZoneById("Pacific Standard Time");

            // List possible invalid times for a 63-minute interval, from 1:59 AM to 3:01 AM
            for (int ctr = 0; ctr < 63; ctr++)
            {
                // Because of assignment, newTime.Kind is also DateTimeKind.Unspecified
                newTime = baseTime.AddMinutes(ctr);
                Console.WriteLine($"{newTime} is invalid: {pstZone.IsInvalidTime(newTime)}");
            }
            // </Snippet9>
        }
    }
}
